packages/programs/src/programs/BedProgram.sol

```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.24;
+
+import { IBaseWorld } from "@latticexyz/world-consumer/src/experimental/WorldConsumer.sol";
+
+import { EntityId } from "@dust/world/src/EntityId.sol";
+
+import { ISleepHook, IWakeupHook } from "@dust/world/src/ProgramInterfaces.sol";
+
+import { DefaultProgram } from "./DefaultProgram.sol";
+
+contract BedProgram is DefaultProgram, ISleepHook, IWakeupHook {
+  constructor(IBaseWorld _world) DefaultProgram(_world) { }
+
+  function onSleep(EntityId caller, EntityId target, bytes memory extraData) external onlyWorld {
+    require(_isAllowed(target, caller), "Only approved callers can sleep in the bed");
+  }
+
+  function onWakeup(EntityId caller, EntityId target, bytes memory extraData) external onlyWorld {
+    require(_isAllowed(target, caller), "Only approved callers can wake up from the bed");
```

Comment:
kinda funny, is there anything you could do if someone changes the program to prevent you from waking up while you're asleep?

Reply:
you couldn't even wait till you're out of energy because the bed keeps your energy level lol

Reply:
> you couldn't even wait till you're out of energy because the bed keeps your energy level lol

you'd have to wait until the force field runs out of energy yea